### PR TITLE
Move test monkeypatches to default.qubit.autograd and default.qubit.tf

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The QNode integration tests currently monkeypatch `default.qubit.autograd` and `default.qubit.tf` to force them to pass the tests. Since backprop is now the new gradient default, it is imperative to move these patches directly to the source code.

**Description of the Change:** Moves the monkeypatching into the source code, removes the `monkeypatch` fixture from the tests.

**Benefits:** Tests pass without monkeypatching

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/devices/default_qubit_autograd.py [hunk 1]
```diff
@@ -116,6 +115,15 @@ class DefaultQubitAutograd(DefaultQubit):
     _roll = staticmethod(np.roll)
     _stack = staticmethod(np.stack)
 
+    @staticmethod
+    def _asarray(array, dtype=None):
+        res = np.asarray(array, dtype=dtype)
+
+        if res.dtype is np.dtype("O"):
+            return np.asarray(np.hstack(array).flatten(), dtype=dtype)
```
Comment (thisac) [inline pennylane/devices/default_qubit_autograd.py:123]: The reason behind using both `np.hstack` and `.flatten()` here to support flattening multdim arrays, right? And the `asarray` is simply to get the `dtype` in as well. Any reason for not using `np.hstack(array).flatten().astype(dtype)` instead (although I would say both work)?

File: tests/tape/interfaces/test_qnode_autograd.py [hunk 2]
```diff
@@ -23,7 +23,7 @@
     "dev_name,diff_method", [
         ["default.qubit", "finite-diff"],
         ["default.qubit", "parameter-shift"],
-        ["default.qubit.autograd", "backprop"],
+        ["default.qubit", "backprop"],
```
Comment (thisac) [inline tests/tape/interfaces/test_qnode_autograd.py:26]: Nice! :slightly_smiling_face:

File: pennylane/tape/qnode.py [hunk 3]
```diff
@@ -284,7 +284,12 @@ def _validate_backprop_method(device, interface):
             if interface in backprop_devices:
                 # TODO: need a better way of passing existing device init options
                 # to a new device?
-                device = qml.device(backprop_devices[interface], wires=device.wires, analytic=True)
+                device = qml.device(
+                    backprop_devices[interface],
+                    wires=device.wires,
+                    shots=device.shots,
```
Comment (thisac) [inline pennylane/tape/qnode.py:290]: Ooh, so this is the fix for #1036. Nice catch (didn't understand how this PR fixed that issue at first)! :octopus:

**Review by thisac — APPROVED** [review]
Looks good @josh146!

File: pennylane/tape/qnode.py (hunk 3 above)
Comment (josh146) [reply pennylane/tape/qnode.py:290]: Yep, changing from `default.qubit.autograd` to `default.qubit` in the pytest mark caused the tests to fail due to the same bug!

File: pennylane/devices/default_qubit_autograd.py (hunk 1 above)
Comment (josh146) [reply pennylane/devices/default_qubit_autograd.py:123]: Oh I like your suggestion better
```suggestion
            return np.hstack(array).flatten().astype(dtype)
```

**Pull request merged**